src/vma/lwip/tcp_out.c

```diff
@@ -852,120 +851,128 @@ tcp_send_empty_ack(struct tcp_pcb *pcb)
+      return;
+    }
 
-		/*Update original buffer*/
-		seg->p->next = NULL;
-		seg->p->len = seg->p->len - lentoqueue;
-		seg->p->tot_len = seg->p->len;
+    /* Copy the data from the original buffer */
+    TCP_DATA_COPY2((char *)p->payload + optlen, (u8_t *)seg->dataptr + lentosend, lentoqueue , &chksum, &chksum_swapped);
 
-		/* Allocate memory for tcp_seg, and fill in fields. */
-		if (NULL == (newseg = tcp_create_segment(pcb, p, 0,  seg->seqno + lentosend, optflags))) {
-			LWIP_DEBUGF(TCP_OUTPUT_DEBUG | 2, ("tcp_output : could not allocate memory for segment\n"));
-			tcp_tx_pbuf_free(pcb, p);
-			return;
-		}
+    /*Update new buffer*/
+    p->tot_len = seg->p->tot_len - lentosend - TCP_HDR;
+    p->next = seg->p->next;
 
-		/*New segment update*/
-		newseg->next = seg->next;
-		newseg->flags = seg->flags;
+    /*Update original buffer*/
+    seg->p->next = NULL;
+    seg->p->len = seg->p->len - lentoqueue;
+    seg->p->tot_len = seg->p->len;
 
-		/* Original segment update */
-		seg->next = newseg;
-		seg->len = seg->p->len - TCP_HDR;
+    /* Allocate memory for tcp_seg, and fill in fields. */
+    if (NULL == (newseg = tcp_create_segment(pcb, p, 0,  seg->seqno + lentosend, optflags))) {
+      LWIP_DEBUGF(TCP_OUTPUT_DEBUG | 2, ("tcp_output : could not allocate memory for segment\n"));
+      tcp_tx_pbuf_free(pcb, p);
+      return;
+    }
 
-		if (pcb->last_unsent == seg) {
-		  pcb->last_unsent = newseg;
-		  pcb->unsent_oversize = oversize;
-		}
+    /*New segment update*/
+    newseg->next = seg->next;
+    newseg->flags = seg->flags;
 
-		/* Set the PSH flag in the last segment that we enqueued. */
-		TCPH_SET_FLAG(newseg->tcphdr, TCP_PSH);
+    /* Original segment update */
+    seg->next = newseg;
+    seg->len = seg->p->len - TCP_HDR;
 
-		/* Update number of buffer to be send */
-		pcb->snd_queuelen++;
-	}
-	else {
-		struct pbuf *pnewhead = seg->p->next;
-		struct pbuf *pnewtail = seg->p;
-		struct pbuf *ptmp = seg->p;
-		u32_t headchainlen = seg->p->len;
-
-		while ((headchainlen + pnewhead->len - TCP_HDR)<= lentosend) {
-			headchainlen += pnewhead->len;
-			pnewtail = pnewhead;
-			pnewhead = pnewhead->next;
-		}
+    if (pcb->last_unsent == seg) {
+      pcb->last_unsent = newseg;
+      pcb->unsent_oversize = oversize;
+    }
+
+    /* Set the PSH flag in the last segment that we enqueued. */
+    TCPH_SET_FLAG(newseg->tcphdr, TCP_PSH);
+
+    /* Update number of buffer to be send */
+    pcb->snd_queuelen++;
+  }
+  else if (seg->p->next) {//Segment with more than one pbuffer, split segment pbuff chain
+    struct pbuf *pnewhead = seg->p->next;
+    struct pbuf *pnewtail = seg->p;
+    struct pbuf *ptmp = seg->p;
+    u32_t headchainlen = seg->p->len;
+
+    while ((headchainlen + pnewhead->len - TCP_HDR)<= lentosend) {
+      headchainlen += pnewhead->len;
+      pnewtail = pnewhead;
+      pnewhead = pnewhead->next;
+    }
```

Comment:
I suggest tracking unexpected flow and return if needed
```
if (pnewhead == NULL) {
    // unexpected situation. Issue message
    return;
}
```
In any case the while loop condition "pnewhead->len" must be protected against NULL address access